### PR TITLE
Refactor base game to use QueryableOrderedSet for Collidables

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

This is a initial step to unify addLater and removeLater between BaseGame and BaseComponent.

I noticed there is some Collidables code that was run in between the operations I want to extract.

However it seems that that can be vastly simplified using the brand new QOS.

@spydon lmk if that makes sense. I believe the only reason it was there was to keep the copy list consistent. now QOS keeps track of all cache lists for you. tests passed and the examples seems to run exactly as intended.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: packages/flame/lib/src/game/base_game.dart [hunk 1]
```diff
@@ -87,13 +87,34 @@ class BaseGame extends Game with FPSCounter {
     _combinedProjector = Projector.compose([camera, viewport]);
   }
 
+  /// This method setps up the OrderedSet instance used by this game, before
+  /// any lifecycle methods happen.
+  ///
+  /// You can return a specific sub-class of OrderedSet, like
+  /// [QueryableOrderedSet] for example, that we use for Collidables.
+  OrderedSet<Component> createOrderedSet() {
```
Comment (luanpotter) [inline packages/flame/lib/src/game/base_game.dart:95]: I always wanted to have this. For very complex games, this allow more easily users to override and return very customized stuff (we do that on Tales for example but we replace the components variable that now is final)

File: packages/flame/lib/src/game/base_game.dart [hunk 2]
```diff
@@ -87,13 +87,34 @@ class BaseGame extends Game with FPSCounter {
...
+  }
+
   /// This method is called for every component added.
   /// It does preparation on a component before any update or render method is called on it.
   ///
   /// You can use this to setup your mixins, pre-calculate stuff on every component, or anything you desire.
   /// By default, this calls the first time resize for every component, so don't forget to call super.preAdd when overriding.
   @mustCallSuper
   void prepare(Component c) {
+    if (c is Collidable) {
```
Comment (luanpotter) [inline packages/flame/lib/src/game/base_game.dart:112]: I am not sure why this check was not consistent with the others. Lmk @spydon if there was a reason.

File: packages/flame/lib/src/game/base_game.dart [hunk 3]
```diff
@@ -201,15 +218,22 @@ class BaseGame extends Game with FPSCounter {
   @override
   @mustCallSuper
   void update(double dt) {
+    _updateComponentList();
```
Comment (luanpotter) [inline packages/flame/lib/src/game/base_game.dart:221]: now, this method is ready to be extracted later

File: packages/flame/lib/src/game/base_game.dart [hunk 4]
```diff
@@ -87,13 +87,34 @@ class BaseGame extends Game with FPSCounter {
...
   }
 
+  /// This method setps up the OrderedSet instance used by this game, before
+  /// any lifecycle methods happen.
+  ///
+  /// You can return a specific sub-class of OrderedSet, like
+  /// [QueryableOrderedSet] for example, that we use for Collidables.
+  OrderedSet<Component> createOrderedSet() {
+    final comparator = Comparing.on<Component>((c) => c.priority);
+    if (this is HasCollidables) {
```
Comment (luanpotter) [inline packages/flame/lib/src/game/base_game.dart:97]: I am still considering going full QOS always. that would make it easier for people to just register/query w/o casting. For now I thought doing this as it's pretty simple and works as almost an A/B test, and we can see if there is any issues with Collidables by using this feature. we then could consider moving to 100% and changing the typedef of the field.

File: packages/flame/lib/src/game/base_game.dart (hunk 2 above)
Comment (spydon) [reply packages/flame/lib/src/game/base_game.dart:112]: That's odd, feels like I must have had a reason, but they look equivalent to me.

**Review by spydon — APPROVED** [review]
Lgtm, this will be very handy!

**Pull request merged**